src/univariate/continuous/uniform.jl

```diff
@@ -109,6 +109,13 @@ function cf(d::Uniform, t::Real)
     cis(v) * (sin(u) / u)
 end
 
+#### Fast path for `loglikelihood`
+
+function loglikelihood(d::Uniform, x::AbstractArray{<:Real})
+    diff = d.b - d.a
+    return all(Base.Fix1(insupport, d), x) ? -length(x) * log(diff) : log(zero(diff))
```

Comment:
Why not
```suggestion
    all_insupport = ChainRulesCore.@ignore_derivatives all(Base.Fix1(insupport, d), x)
    return all_insupport ? -length(x) * log(diff) : log(zero(diff))
```
? Then the AD only needs to differentiate `log` and the ternary, and a custom rule is not needed (Zygote will still be a little slower since hand-derived rules are almost always faster than Zygote's generated codes, especially if there's control flow).

Reply:
I considered something similar first 😄

Since the code that still has to be differentiated is basically like `logpdf` we  should include a custom rule though - we do for `logpdf` and hence it seems we should also for `loglikelihood` which is an (at least as) integral part of e.g. Turing. The rule for `logpdf` was included in DistributionsAD (well, still is but there's a PR that fixes it) - as type piracy, of course, but it is only possible to fix it without performance regression if we define it in Distributions instead.

Reply:
Oh, and even more importantly it's not clear that AD returns the desired derivatives for the boundaries and outside of the support.

Reply:
I guess if this is part of transferring from DistributionsAD to here, then it should be here.



> Oh, and even more importantly it's not clear that AD returns the desired derivatives for the boundaries and outside of the support.

Why would it not? As far as I can tell the rules included here just manually do what the AD would do.

Reply:
It's exactly what Zygote would do. I'm not sure about general AD though, eg. also FiniteDifferences returns `NaN` instead in these cases.

Reply:
> FiniteDifferences returns `NaN` instead in these cases.

FD is not an AD. 🙂

Reply:
I'm not worried about FD specifically, just worried that the same NaN issues would occur with AD. But maybe that can't happen?

Reply:
Ah, there's another problem and something I forgot to fix in https://github.com/JuliaStats/Distributions.jl/pull/1459: We want to return 0 also at the boundary. This is a bug in the current implementation, and also not what you get from Zygote without rule.

Reply:
I fixed the bugs, added more tests, and updated the benchmarks above.

Reply:
Ah yes, I remember reading that discussion. Interesting, I would have thought returning the same thing at the boundary as for the rest of the support would be best, but from the perspective of someone maximizing the likelihood over the parameters starting with bounds that include all `x`, this would result in the maximum being achieved with a non-zero gradient (the bounds being the extrema of `x`), the next step being to shorten the bounds to not include all `x`. So yes there's a case to be made for returning 0 at the boundary. Still not 100% convinced this is good enough of a reason to override the default AD behavior.

Reply:
Yeah, it's clearly not the only choice.

To recap, if we neglect `x` values outside the support completely, then we have to study `f(a) = -log(b - a)`, defined on `(-Inf, x]` (`b` and `x <= b` fixed), and `g(b) = -log(b - a)`, defined on `[x, Inf)` (`a` and `x >= a` fixed). Clearly, for `a < x` we have `f'(a) = 1/ (b - a) > 0` and for `b > x` we have `g'(b) = 1 / (a - b) < 0`.

A straightforward, and I think reasonable, way to define "derivatives" at the non-differentiable points `a = x` and `b = x` would to use the limit from the left and the right, respectively - as also mentioned as one approach in the CR docs. Ie., to set `f'(x) = 1 / (b - a)` and `g'(x) = 1 / (a - b)`. However, this means that the derivatives at the global maxima of `f` and `g` would be non-zero - and hence not satisfy the property, useful in practice and also mentioned in the CR docs, that derivatives at local maxima and minima should be zero.

Thus another approach would be to set `f'(x) = g'(x) = 0`. This is what we wanted to do and what this PR implements. It satisfy the "zero at extrema" property and hence I still think it is the more useful definition in practice as eg. it won't push you over the edge in a gradient descent algorithm. If one uses the convention to set the derivatives of `f` and `g` to 0 if `x` is not in the support, then it even corresponds to the limits from the right and the left, respectively. And I think the increased usefulness in practice outweighs the inconvenience of having to define custom rules.

(Another alternative would be to error or set the derivative to `NaN` but I agree with the CR docs that this is the least useful approach.)

Reply:
The last comment made me even more certain that we should define a rule in any case: Regardless of the convention we use, since there is no mathematically correct definition of these derivatives we don't have any guarantees that AD will necessarily adhere to our convention. Similar to the `abs(x)` example in the CR docs, slightly different implementations of the primal or internals of some AD backend may yield different derivatives following a different convention.

Reply:
And my comment made me realize that, of course, we have to distinguish between `x = a` and `x = b`. And hence the derivatives don't follow our desired convention yet 😂

Reply:
Something we didn't use as a criterion is that the derivative we return could well be an convenient element of the set of sub gradients.

Reply:
That would suggest `0` and not `NaN` for `abs` for example.

Reply:
I actually considered it and mentioned it in the original PR, I just didn't mention it above because I hadn't thought about it carefully and noticed that only `f` is a convex function.

I'm used to definitions of subderivatives in open sets, but anyway in the case of `f` a subderivative `c` at `x` would have to satisfy `f(a) - f(x) >= c (a - x)` for all `a <= x`. We have `f(a) - f(x) = log(b - x) - log(b - a) < 0` for `a < x`, and hence 0 would not be a subderivative (if we assume it makes sense to talk about subderivatives at the boundary). However, since `f(a) - f(x) = log((b - x) / (b - a)) >= 1 - (b - a) / (b - x) = (a - x) / (b - x)` for all `a <= x` (logarithm inequality), `1 / (b - x)` is a subderivative at `x` (and `[1 / (b - x), Inf)` is a subset of subderivatives) [assuming `b != x`].

Similarly, for `g` (which is not convex, another property that's often assumed) a subderivative `c` at `x` would have to satisfy `g(b) - g(x) = log(x - a) - log(b - a) >= c(b - x)` for all `b >= x`. Similarly, `log(x - a) - log(b - a) < 0` for `b > x` and hence 0 is not a subderivative. We have `g(b) - g(x) >= 1 - (b - a) / (x - a) = (b - x) / (a - x)` for all `x >= b`, and thus `1 / (a - x)` is a subderivative at `x` (and `(-Inf, 1 / (a - x)]` is a set of subderivatives) [assuming `a != x`].

So unsurprisingly the subderivatives at the boundary `x` would include the limits of `f'` and `g'` from the left and right side, respectively. Whereas 0 is no subderivative.

The set of subderivatives also provides only a rough guidance - for the (convex!) function `abs` the set of subderivatives at 0 is [-1, 1], so it's not a clear argument for choosing 0 but mainly not for not choosing `NaN`. Which is a bad idea anyway usually.

Reply:
I'm not certain subderivatives are reasonable concepts for non-convex functions. e.g. the geometric picture of the subderivative is the set of slopes of all tangent lines touching the curve at the desired point such that the tangent lines are completely beneath the curve, which is not possible here, since we have a step function at the bounds. But the subderivative is just the special case for convex functions of the more general "use the derivatives at either side of the point to define a set", in which case we basically have `[0, 1/(b-a)]` being the set of reasonable options. I think we've agreed `NaN` is a bad option. `Inf` is almost as bad. @devmotion it sounds like you still favor 0, while AD would default to `1/(b-a)`. From the perspective of maximizing the log-likelihood, I agree that `0` makes the most sense, but this set also includes what the AD would do by default, so I still am not convinced we should override that.

> The last comment made me even more certain that we should define a rule in any case: Regardless of the convention we use, since there is no mathematically correct definition of these derivatives we don't have any guarantees that AD will necessarily adhere to our convention. Similar to the `abs(x)` example in the CR docs, slightly different implementations of the primal or internals of some AD backend may yield different derivatives following a different convention.

By this reasoning, every conditional expression in Julia needs a custom rule where someone has injected semantic information about what constitutes the most reasonable derivative for optimization purposes. But AD is widely used, and that doesn't seem to be necessary for most functions. But for extremely common functions, and especially those that have discontinuities at a region that is likely to be the target of optimization, these are useful discussions to have. The question is, is anyone going to (reasonably) try to maximize the likelihood of a uniform distribution using the gradient, when the MLE is known in closed form? Will this be regularly done? And by this same reasoning, must we add custom rules for all distributions with bounded support to special-case the bounds? My concern is that this seems like over-engineering.

Reply:
> And by this same reasoning, must we add custom rules for all distributions with bounded support to special-case the bounds?

To clarify, I'm referring to distributions whose support is determined by their parameters. e.g. `AffineDistribution` or `Pareto`.

Reply:
I think the discussion of sub gradients points in the right direction: It is difficult to come with a rule and guarantee what derivative is returned at these points, but it might be possible to give a guarantee that we always return an (arbitrary) element of a well-described set, and if something is differentiable, that set should contain only one point.

Reply:
> The question is, is anyone going to (reasonably) try to maximize the likelihood of a uniform distribution using the gradient, when the MLE is known in closed form?

A stray uniform might show up somewhere in a more complicated composite likelihood from a PPL model?

Reply:
Yeah, I'm still in favour of 0s. There's no correct choice and hence I think we should choose out of all somewhat reasonable values the ones that are most useful in practice.

There seems to also be a misunderstanding about why I think that a rule should be defined in any case. My point is, if a correct derivative exists and AD returns something else, then it's a bug of the AD - we might fix it by adding a CR definition but there's no need to define a rule in advance for correctness (might want to do it for performance, mutation issues, etc. though). However, if there's no correct derivative, then AD is allowed to do follow different conventions, possibly depending on the implementation, but it's not a bug of AD if it's something else than what we prefer. So a rule is the only way we can ensure that we always get the values we want to get. But it's not relevant if the code contains `if` statements, this argument is solely based on the non-existence of the derivative.

The rule here exists both for performance and correctness reasons. Initially, it was defined in DistributionsAD (and used NaN 🙈), and there's still one eg for Tracker (IIRC there was also one for ForwardDiff at some point, not sure what it returns, maybe `NaN` outside the support?). In any case its removal would lead to performance regressions and mean we don't have control over the values at the boundary and outside the support anymore.

It's not a completely unusual example either, actually for multiple distributions we already define CR definitions, both for correctness and performance: https://github.com/JuliaStats/StatsFuns.jl/blob/master/src/chainrules.jl

So I think the main question is, do we stick with the 0s as discussed previously? Or do we use something else. If this is decided, I can finish the PR.